embassy-rp/src/rom_data.rs

```diff
@@ -114,43 +90,74 @@ macro_rules! declare_rom_function {
             /// Retrieve a function pointer.
             #[cfg(not(feature = "rom-func-cache"))]
-            pub fn ptr() -> unsafe extern "C" fn( $($argname: $ty),* ) -> $ret {
+            pub fn ptr() -> $( $maybe_unsafe )? extern "C" fn( $($argname: $ty),* ) -> $ret {
                 let p: *const u32 = $lookup;
                 unsafe {
-                    let func : unsafe extern "C" fn( $($argname: $ty),* ) -> $ret = core::mem::transmute(p);
+                    let func : $( $maybe_unsafe )? extern "C" fn( $($argname: $ty),* ) -> $ret
+                        = core::mem::transmute(p);
                     func
                 }
             }
 
-            /// Retrieve a function pointer.
             #[cfg(feature = "rom-func-cache")]
-            pub fn ptr() -> unsafe extern "C" fn( $($argname: $ty),* ) -> $ret {
-                use core::sync::atomic::{AtomicU16, Ordering};
-
-                // All pointers in the ROM fit in 16 bits, so we don't need a
-                // full width word to store the cached value.
-                static CACHED_PTR: AtomicU16 = AtomicU16::new(0);
-                // This is safe because the lookup will always resolve
-                // to the same value.  So even if an interrupt or another
-                // core starts at the same time, it just repeats some
-                // work and eventually writes back the correct value.
-                let p: *const u32 = match CACHED_PTR.load(Ordering::Relaxed) {
-                    0 => {
-                        let raw: *const u32 = $lookup;
-                        CACHED_PTR.store(raw as u16, Ordering::Relaxed);
-                        raw
-                    },
-                    val => val as *const u32,
-                };
-                unsafe {
-                    let func : unsafe extern "C" fn( $($argname: $ty),* ) -> $ret = core::mem::transmute(p);
-                    func
+            paste::paste! {
+                use core::sync::atomic::{compiler_fence, Ordering};
+
+                // unlike rp2040-hal we store a full word, containing the full function pointer.
+                // rp2040-hal saves two bytes by storing only the rom offset, at the cost of
+                // having to do an indirection and an atomic operation on every rom call.
+                extern "C" {
+                    static mut [<$name _cache>]: $( $maybe_unsafe )? extern "C" fn( $($argname: $ty),* ) -> $ret;
+                }
+
+                #[link_section = concat!(".text.rom_func_cache.", stringify!([<$name _cache>]), ".trampoline")]
+                extern "C" fn lookup() -> *const u32 {
+                    $lookup
+                }
+
+                core::arch::global_asm!{
+                    r#" .section .data.rom_func_cache.{cache}
+                        .type {cache}, %object
+                        .align 2
+                        {cache}:
+                            .word rom_func_cache.{cache}.trampoline
```

Comment:
why is it needed to define this in ASM instead of Rust? Doing it in rust would allow letting rustc mangle it, which would allow getting rid of `paste!`.

Reply:
I mean the variable, not the function. I think it should be possible to define the variable as a regular Rust `static mut` and then have the ASM use it?